interfaces/hotplug/udevadm.go

```diff
@@ -0,0 +1,110 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package hotplug
+
+import (
+	"bufio"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+var udevadmBin = `/sbin/udevadm`
+var udevadmRe = regexp.MustCompile("^([A-Z]): (.*)$")
+
+// RunUdevadm enumerates all devices by parsing 'udevadm info -e' command output and reports them
+// via devices channel. The devices channel gets closed to indicate that all devices were processed.
+// Parsing errors are reported to parseErrors channel. Fatal errors encountered when starting udevadm
+// are reported by the error return value.
+func RunUdevadm(devices chan<- *HotplugDeviceInfo, parseErrors chan<- error) error {
+	cmd := exec.Command(udevadmBin, "info", "-e")
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		close(devices)
+		return err
+	}
+
+	rd := bufio.NewScanner(stdout)
+	if err := cmd.Start(); err != nil {
+		close(devices)
+		return err
+	}
+
+	go func() {
+		defer close(devices)
+
+		env := make(map[string]string)
+		for rd.Scan() {
+			line := rd.Text()
+			if line == "" {
+				if len(env) > 0 {
+					outputDevice(env, devices, parseErrors)
+					env = make(map[string]string)
+				}
+				continue
+			}
+			match := udevadmRe.FindStringSubmatch(line)
+			if len(match) > 0 {
+				// possible udevadm line prefixes are:
+				//  'P' - device path (e.g /devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.0/ttyUSB0)
+				//  'N' - device node (e.g. "/dev/ttyUSB0")
+				//  'L' - devlink priority (unclear)
+				//  'S' - devlinks (e.g. serial/by-path/pci-0000:00:14.0-usb-0:2:1.0-port0)
+				//  'E' - property (key=value format)
+				// We are only interested in 'E' properties as they carry all the interesting data,
```

Comment:
I'm not sure we should just ignore `P:` lines. Those are consistently included for each device and I'd think about using them as markers for device section/block start followed by an empty line which marks the end of device block.

Reply:
Ok, added extra check for presence of "P: " to mark start of a device block.